Add tests to aid in debugging dynamo + aot_autograd dedup

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #90370
* __->__ #90130
* #90129

This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
wip

wip

wip

WIP

Useful tests

lint

Extend test to cover failure reason

fix

[ghstack-poisoned]

voznesenskym commented on test/functorch/test_aotdispatch.py:
```diff
@@ -1070,8 +1070,45 @@ def forward(self, x, y):
+        )
+
+    @patch('functorch._src.aot_autograd.AOT_COUNTER', new_callable=itertools.count)
+    @patch("functorch._src.config.debug_assert", True)
```
bad merge?

voznesenskym committed:
Update on "Add two useful tests to aid in debugging dynamo + aot_autograd dedup"


This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

albanD commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -281,6 +283,72 @@ def capturing_fn(gm, inputs):
+
+        def _outs_and_grads(fn, graph_inps, inps):
+            outs = fn(*graph_inps)
+            for out in pytree.tree_flatten(outs)[0]:
+                if isinstance(out, torch.Tensor) and out.requires_grad:
+                    out.sum().backward(retain_graph=True)
+            grads = [inp.grad for inp in pytree.tree_flatten(inps)[0]]
+            for inp in pytree.tree_flatten(inps)[0]:
+                inp.grad = None
```
This is a simpler way to do this
```suggestion
           flat_outs = pytree.tree_flatten_only(torch.Tensor, outs)[0]
           flat_inps = pytree.tree_flatten_only(torch.Tensor, inps)[0]
           grads = torch.autograd.grad((o for o in flat_outs if o.requires_grad), flat_inps)
           return outs, grads
```

ezyang replied:
Also, don't we already have this helper function defined, why do we need another copy?

ezyang commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -281,6 +283,72 @@ def capturing_fn(gm, inputs):
+
+        x = torch.randn(3, 3, requires_grad=True)
+        y = torch.randn(3, 3, requires_grad=True)
+        z = torch.randn(3, 3, requires_grad=False)
```
nit: it would have been better to have more descriptive name for y and z; e.g., y_true and y_false or y_grad and y_nograd, to make it more clear which ones is requires grad True and which one is False.

ezyang commented:
I mentioned this in earlier review; these tests are straight up copy pastes of already existing tests in the test suite. Lines of code in the test suite are a liability, not an asset, since the more lines of test code you have, the harder it is to refactor APIs (since you now have to update all the test code.) Please make more effort to deduplicate these tests.

voznesenskym commented:
> I mentioned this in earlier review; these tests are straight up copy pastes of already existing tests in the test suite. 

One is, and one is different enough and needs some dynamo test imports that we'd probably need to keep it. But for the aot_dispatch test, 100% agreed. 

> Lines of code in the test suite are a liability, not an asset, since the more lines of test code you have, the harder it is to refactor APIs (since you now have to update all the test code.) Please make more effort to deduplicate these tests.

Of course :) I agree! Higher level - does the reasoning in the tests for why we have these two seem sound? I was hoping more to get a take from you on that.

ezyang commented:
The test itself is fine. I'm imagining a new type of test which is done in test/dynamo, but the model in question is simple enough that you can also run it through directly with aot_module_simplified. Or maybe something like OpInfo but ProgramInfo, where we have sample programs that can be run through various levels of the stack.

voznesenskym committed:
Update on "Add two useful tests to aid in debugging dynamo + aot_autograd dedup"


This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add two useful tests to aid in debugging dynamo + aot_autograd dedup"


This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

voznesenskym committed:
Update on "Add tests to aid in debugging dynamo + aot_autograd dedup"


This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
i'm unsubsubscribing, rerequest review when ready

voznesenskym committed:
Update on "Add tests to aid in debugging dynamo + aot_autograd dedup"


This adds a pair of equivalent tests, both of which invoke aot_autograd with the same inputs, more or less. One through dynamo, one directly. Where the direct call fails, dynamo passes due to guards. This pair of tests protects that behavior and sets us up for future refactors. 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen chunyuan-w XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]
